utils/bundle_miner.c

```diff
@@ -0,0 +1,157 @@
+  uint64_t optimal_index;
+  uint64_t count;
+  double probability;
+} bundle_miner_ctx_t;
+
+/**
+ * Private functions
+ */
+
+static void *bundle_miner_mine_routine(void *const param) {
+  Kerl kerl;
+  double probability = 0.0;
+  trit_t candidate[HASH_LENGTH_TRIT];
+  byte_t candidate_normalized[NORMALIZED_BUNDLE_LENGTH];
+  byte_t candidate_normalized_max[NORMALIZED_BUNDLE_LENGTH];
+  bundle_miner_ctx_t *ctx = (bundle_miner_ctx_t *)param;
+
+  kerl_init(&kerl);
+
+  for (size_t i = 0; i < ctx->count; i++) {
+    long_to_trits(ctx->index, ctx->essence + OBSOLETE_TAG_OFFSET);
+
+    kerl_reset(&kerl);
+    kerl_absorb(&kerl, ctx->essence, ctx->essence_length);
+    kerl_squeeze(&kerl, candidate, HASH_LENGTH_TRIT);
+
+    normalize_hash(candidate, candidate_normalized);
+
+    if (normalized_hash_is_secure(candidate_normalized, ctx->security * NORMALIZED_FRAGMENT_LENGTH)) {
+      bundle_miner_normalized_bundle_max(ctx->bundle_normalized_max, candidate_normalized, candidate_normalized_max,
+                                         ctx->security * NORMALIZED_FRAGMENT_LENGTH);
+      probability = bundle_miner_probability_of_losing(candidate_normalized_max, ctx->security);
+      if (probability < ctx->probability) {
+        ctx->probability = probability;
+        ctx->optimal_index = ctx->index;
+      }
+    }
+
+    ctx->index += 1;
+  }
+
+  return NULL;
+}
+
+/**
+ * Public functions
+ */
+
+double bundle_miner_probability_of_losing(byte_t const *const normalized_hash, uint8_t const security) {
+  double p = 0.0;
+  double pi = 0.0;
+
+  for (size_t i = 0; i < security * NORMALIZED_FRAGMENT_LENGTH; i++) {
+    pi = 1.0 - ((double)(TRYTE_VALUE_MAX - normalized_hash[i]) / (double)(TRYTE_VALUE_MAX - TRYTE_VALUE_MIN));
+    if (pi > 0.0) {
+      if (p == 0.0) {
+        p = 1.0;
+      }
+      p *= pi;
+    }
+  }
+
+  return p;
+}
+
+void bundle_miner_normalized_bundle_max(byte_t const *const lhs, byte_t const *const rhs, byte_t *const max,
+                                        size_t const length) {
+  for (size_t i = 0; i < length; i++) {
+    max[i] = MAX(lhs[i], rhs[i]);
+  }
+}
+
+retcode_t bundle_miner_mine(byte_t const *const bundle_normalized_max, uint8_t const security,
+                            trit_t const *const essence, size_t const essence_length, uint32_t const count,
+                            uint8_t const nprocs, uint64_t *const index) {
+  retcode_t ret = RC_OK;
+  uint8_t procs = nprocs == 0 ? system_cpu_available() : nprocs;
+  uint32_t rounded_count = count + (procs - count % procs);
+  bundle_miner_ctx_t *ctxs = (bundle_miner_ctx_t *)malloc(sizeof(bundle_miner_ctx_t) * procs);
+  thread_handle_t *threads = (thread_handle_t *)malloc(sizeof(thread_handle_t) * procs);
+  uint64_t start_index = 0;
+  double probability = 1.0;
+
+  if (bundle_normalized_max == NULL || essence == NULL || index == NULL) {
+    return RC_NULL_PARAM;
+  }
+
+  if (security > 3 || essence_length % HASH_LENGTH_TRIT != 0) {
```

Comment:
doesn't essence always have the same size?

Reply:
It depends on the bundle size, but this check is only a safety measure to avoid kerl to crash.